bylaws/008-psr-evolution.md

```diff
@@ -0,0 +1,73 @@
+
+This document defines a guideline to be followed in updating PSR interfaces, in a way that is not breaking in regard to behavior for end users, and with an appropriate upgrade path for the packages.
+
+## New releases
+
+A new minor release of a PHP-FIG package containing interfaces for a PSR MUST follow these rules:
+ * the new release MUST follow [Semantic Versioning](https://semver.org/) rules;
+ * the PSR behavior MUST NOT be altered;
+ * the packages that implement the interfaces or users that are consuming them MUST NOT suffer breaking changes;
+ * the PHP version constraint of the PHP-FIG package MAY be altered to require newer language features that would aid cross-compatibility;
+ * the PHP version constraint of the PHP-FIG package MUST NOT be altered to use newer language features that would create cross-compatibility issues.
+ 
+A new major release of a PHP-FIG package containing interfaces for a PSR MUST follow the same rules, with this exception:
+ * the new major version of the package MAY contain breaking changes if the implementing packages have a reasonable upgrading path, like the possibility of releasing a cross-compatible implementation with the previous releases.
+ * the new major version of the package MAY refer to a new, superseding PSR.
+
+### Workflow
+
+Since releasing new versions of the interfaces MUST NOT alter the PSR in its behavior, those releases can be voted in with the same process as errata changes. The new releases MUST be declared and embedded in the PSR document, with eventual indications on the upgrade path in the meta document.
+
+### Practical example
+
+A common case for an upgrade in the interfaces is to add parameters and return types, since they are a new language feature introduced by PHP 7, and many PSR interfaces predates that release. In the next paragraph, we'll try to use PSR-3 as an example to show how a PSR interface should be updated.
+
+#### PSR-3: the interface
+
+PSR-3 is released with the [`psr/log` package](https://packagist.org/packages/psr/log) and it contains the `LoggerInterface`, which has this method:
+```php
+/**
+ * Logs with an arbitrary level.
+ *
+ * @param mixed  $level
+ * @param string $message
+ * @param array  $context
+ *
+ * @return void
+ */
+public function log($level, $message, array $context = array());
+```
+This method could be updated with a new minor release that adds the argument type for `$message`:
+```php
+public function log($level, string $message, array $context = []);
+```
+This change would be technically a breaking change but, thanks to the [limited contravariance possible in PHP 7.2](https://wiki.php.net/rfc/parameter-no-type-variance), we can avoid that. This means that just by requiring `"php": "^7.2"` in the `prs/log` `composer.json`, we could tag this change as a minor release, and have all the implementors be automatically cross-compatible, provided that they declare `"psr/log": "^1.0"` (or equivalent) as a constraint, which they should.
```

Comment:
Indeed, this would not be a breaking change for implementing library.
But, as I was explaining on the thread, it will be a breaking change for project and libraries that calls the `log` method.
Even if the docblock specified what message parameter type is, there was no strong enforcement. After a minor release, there will be a language runtime enforcement.

Just as return types can be reduced for callers but cannot be reduced for implementors, similarly, parameter types can be reduced for implementors but cannot be reduced for callers.

Reply:
That's not true: the fact that the interface is requiring a certain type is not enforced until the library adds the same type too, thanks to PHP 7.2. At that point, it should be up to the libraries to decide if they want to adhere strictly to the interface, add the types and create a breaking change or not.

Due to LSP, implementing libraries are free to accept MORE types than the one required by the interface.

Reply:
I agree that when adding types to parameters there is no breaking changing on the implementing libraries.
I argue that there would be a breaking change on the consuming libraries.

Reply:
That depends totally on the implementing libraries. As highlighted in https://github.com/php-fig/fig-standards/pull/1195#discussion_r334586301, for Monolog specifically they could choose to not enforce `string` due to stringable objects being accepted.